File: packages/algorithm/Leetcode/hot100/数据结构/二叉树/从前序与中序遍历构建二叉树.ts
import {TreeNode} from './types'

// 使用slice，简洁！
function buildTree(preorder: number[], inorder: number[]): TreeNode | null {
    if (preorder.length === 0 || inorder.length === 0) {
        return null;
    }

    const rootVal = preorder[0];
    const root = new TreeNode(rootVal);

    const mid = inorder.indexOf(rootVal);

    // inorder用slice的索引规则跳过了rootVal
    root.left = buildTree(preorder.slice(1, mid + 1), inorder.slice(0, mid));
    root.right = buildTree(preorder.slice(mid + 1), inorder.slice(mid + 1))

    return root
};

File: packages/program-design/alog/数据结构/二叉树/从前序与中序遍历构建二叉树.py
# Definition for a binary tree node.
from typing import List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
    
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        valueToIndex = {}
        for i, n in enumerate(inorder):
            valueToIndex[n] = i
        
        def build(preStart, preEnd, inStart, inEnd):
            if preStart > preEnd:
                return None
            rootVal = preorder[preStart]
            index = valueToIndex.get(rootVal)
            leftSize = index - inStart
            
            root = TreeNode(rootVal)

            root.left = build(preStart + 1, preStart + leftSize, inStart, index - 1)
            root.right = build(preStart + leftSize + 1, preEnd, index + 1, inEnd)
            
            return root
        
        return build(0, len(preorder) - 1, 0, len(inorder) - 1)


        